File: src/lib/scoring.ts
/**
 * Reconstruct each campaign's budget state timeline from change events.
 *
 * The export lists only *changes*, so the state between two rows has to be
 * inferred. Three things make that non-obvious:
 *
 * 1. Rows are written newest-first, so rows sharing a minute are also
 *    newest-first and must be reversed before walking the machine. Sorting on
 *    timestamp alone silently preserves the wrong intra-minute order -- on the
 *    reference file that costs 64 campaign-hours and manufactures 14 phantom
 *    inconsistencies.
 * 2. Delivery state (Paused) overlays budget state. A paused campaign forgoes
 *    nothing to its budget, so paused minutes are excluded from the loss-eligible
 *    total and from the in-budget denominator that sets the spend rate.
 * 3. De-duplication during export can drop an intermediate transition, leaving a
 *    row whose `From` disagrees with the running state. We trust the row over the
 *    inference and place the implied transition at the midpoint of the gap.
 */

import { parseMoney } from './ingest';
import type { Event } from './ingest';

export const IN = 0;
export const OOB = 1;
export const PAUSED = 2;
export const NA = 3;

export const MINUTES_PER_DAY = 1440;
export const DEFAULT_MERGE_GAP_MIN = 5;

export interface ChainBreak {
  atMin: number;
  expectedFrom: string;
  sawFrom: string;
  ambiguityMin: number;
}

export interface Episode {
  index: number;
  startMin: number;
  endMin: number;
  rawMin: number;
  activeMin: number; // raw minus any overlapping pause
}

export type BudgetSource = 'daily_budget_event' | 'budget_rule' | 'perf_report' | 'unknown';

export interface BudgetInfo {
  value: number | null;
  source: BudgetSource;
  timeWeighted: number | null;
  changes: number;
}

export interface LostOpportunity {
  rateReliable: boolean;
  spendRatePerHour: number | null;
  lostSpend: number | null;
  lostSales: number | null;
  capped: boolean;
  budgetUsed: number;
}

export type Track = [state: number, start: number, end: number][];

export interface CampaignDay {
  campaign: string;
  dateKey: string;
  t0: number;
  t1: number;

  inMin: number;
  oobMin: number; // loss-eligible: pause and out-of-window already removed
  pausedMin: number;
  naMin: number;
  oobMinRaw: number; // what the budget machine alone says
  postResetOobMin: number;

  episodes: Episode[];
  episodesRaw: number;
  episodesMerged: number;

  firstOobMin: number | null;
  lastRecoveryMin: number | null;
  openedOob: boolean;
  closedOob: boolean;

  hourlyOob: number[];
  hourlyPaused: number[];
  hourlyNa: number[];
  track: Track;

  budget: BudgetInfo;
  chainBreaks: ChainBreak[];
  oobUncertaintyMin: number;
  confidence: 'clean' | 'repaired' | 'partial_day';

  // filled in by metrics.ts / perfjoin.ts
  severity: number;
  diagnosis: string;
  lost: LostOpportunity | null;
  perf: unknown;
}

export const eligibleMin = (d: CampaignDay) => d.t1 - d.t0;

/** Eligible minutes where the campaign could actually have spent. */
export const activeMin = (d: CampaignDay) => eligibleMin(d) - d.pausedMin;

export const oobShare = (d: CampaignDay) => (activeMin(d) > 0 ? d.oobMin / activeMin(d) : 0);

export const inHours = (d: CampaignDay) => d.inMin / 60;
export const oobHours = (d: CampaignDay) => d.oobMin / 60;
export const pausedHours = (d: CampaignDay) => d.pausedMin / 60;

/** Chronological order. Descending source index reverses the newest-first file. */
function bySortKey(a: Event, b: Event): number {
  return a.minute - b.minute || b.sourceIndex - a.sourceIndex;
}

type Span = [state: string, start: number, end: number];

/** Turn a two-state event stream into contiguous spans over [t0, t1). */
function walk(
  events: Event[],
  t0: number,
  t1: number,
  initial: string,
): { spans: Span[]; breaks: ChainBreak[] } {
  const spans: Span[] = [];
  const breaks: ChainBreak[] = [];
  let state = initial;
  let prev = t0;

  for (const e of events) {
    const m = e.minute > prev ? e.minute : prev;
    if (e.fromVal !== state) {
      // A transition went missing. The row is observation, the running
      // state is inference, so trust the row and split the difference.
      const mid = Math.floor((prev + m) / 2);
      if (mid > prev) spans.push([state, prev, mid]);
      if (m > mid) spans.push([e.fromVal, mid, m]);
      breaks.push({ atMin: m, expectedFrom: state, sawFrom: e.fromVal, ambiguityMin: m - prev });
    } else if (m > prev) {
      spans.push([state, prev, m]);
    }
    state = e.toVal;
    prev = m;
  }

  if (t1 > prev) spans.push([state, prev, t1]);
  return { spans, breaks };
}

/** Recover the daily budget, which can change during the day. */
function budgetTimeline(
  events: Event[],
  ruleEvents: Event[],
  t0: number,
  t1: number,
): BudgetInfo {
  const amounts = [...events].sort(bySortKey);
  if (amounts.length) {
    const spans: [number | null, number, number][] = [];
    let value = amounts[0].fromNum;
    let prev = t0;
    for (const e of amounts) {
      const m = Math.max(e.minute, prev);
      if (m > prev) spans.push([value, prev, m]);
      value = e.toNum;
      prev = m;
    }
    if (t1 > prev) spans.push([value, prev, t1]);

    let weighted = 0;
    let covered = 0;
    for (const [v, a, b] of spans) {
      if (v === null) continue;
      weighted += v * (b - a);
      covered += b - a;
    }
    return {
      value: amounts[amounts.length - 1].toNum,
      source: 'daily_budget_event',
      timeWeighted: covered ? weighted / covered : null,
      changes: amounts.length,
    };
  }

  for (const e of ruleEvents) {
    const amount = parseMoney(e.fromVal, e.toVal);
    if (amount !== null) {
      return { value: amount, source: 'budget_rule', timeWeighted: amount, changes: 0 };
    }
  }

  return { value: null, source: 'unknown', timeWeighted: null, changes: 0 };
}

function countIn(flat: Uint8Array, value: number, from: number, to: number): number {
  let n = 0;
  for (let i = from; i < to; i++) if (flat[i] === value) n += 1;
  return n;
}

/** Score one campaign for one day. Returns null if budget state is unknowable. */
export function scoreCampaignDay(
  campaign: string,
  dateKey: string,
  events: Event[],
  dayEndMin = MINUTES_PER_DAY,
  mergeGapMin = DEFAULT_MERGE_GAP_MIN,
): CampaignDay | null {
  let budgetEvents = events.filter((e) => e.machine === 'budget').sort(bySortKey);
  if (!budgetEvents.length) return null; // never impute a state we did not observe

  let deliveryEvents = events.filter((e) => e.machine === 'delivery').sort(bySortKey);
  const created = events.filter((e) => e.machine === 'created');

  // Eligible window. A campaign created at 07:02 is scored over the remaining
  // 16.97h, not a full day -- but only if no budget event precedes creation.
  const t1 = Math.min(dayEndMin, MINUTES_PER_DAY);
  let t0 = 0;
  if (created.length) {
    const birth = Math.min(...created.map((e) => e.minute));
    if (birth <= budgetEvents[0].minute && birth < t1) t0 = birth;
  }
  budgetEvents = budgetEvents.filter((e) => e.minute >= t0);
  if (!budgetEvents.length) return null;
  deliveryEvents = deliveryEvents.filter((e) => e.minute >= t0);

  const budgetWalk = walk(budgetEvents, t0, t1, budgetEvents[0].fromVal);
  const budgetSpans = budgetWalk.spans;
  const breaks = budgetWalk.breaks;

  const deliveryInitial = deliveryEvents.length ? deliveryEvents[0].fromVal : 'Delivering';
  const deliverySpans = walk(deliveryEvents, t0, t1, deliveryInitial).spans;

  // Flatten: not-eligible > paused > out of budget > in budget.
  const flat = new Uint8Array(MINUTES_PER_DAY).fill(NA);
  for (const [state, a, b] of budgetSpans) {
    flat.fill(state === 'Out of budget' ? OOB : IN, a, b);
  }
  for (const [state, a, b] of deliverySpans) {
    if (state === 'Paused') flat.fill(PAUSED, a, b);
  }

  const hourly = (value: number) =>
    Array.from({ length: 24 }, (_, h) => countIn(flat, value, h * 60, (h + 1) * 60));

  const oobSpans: [number, number][] = budgetSpans
    .filter(([s]) => s === 'Out of budget')
    .map(([, a, b]) => [a, b]);

  // Amazon can release a sliver of budget that is consumed within the same
  // minute, producing zero-length recoveries. Those are pacing noise, not
  // genuine outages, so also report a merged count.
  const merged: [number, number][] = [];
  for (const [a, b] of oobSpans) {
    const last = merged[merged.length - 1];
    if (last && a - last[1] < mergeGapMin) last[1] = b;
    else merged.push([a, b]);
  }

  // Run-length encode for the report; spans are contiguous and tile the day.
  const track: Track = [];
  let runState = flat[0];
  let runStart = 0;
  for (let m = 1; m < MINUTES_PER_DAY; m++) {
    if (flat[m] !== runState) {
      track.push([runState, runStart, m]);
      runState = flat[m];
      runStart = m;
    }
  }
  track.push([runState, runStart, MINUTES_PER_DAY]);

  const naMin = countIn(flat, NA, 0, MINUTES_PER_DAY);

  return {
    campaign,
    dateKey,
    t0,
    t1,
    inMin: countIn(flat, IN, 0, MINUTES_PER_DAY),
    oobMin: countIn(flat, OOB, 0, MINUTES_PER_DAY),
    pausedMin: countIn(flat, PAUSED, 0, MINUTES_PER_DAY),
    naMin,
    oobMinRaw: oobSpans.reduce((s, [a, b]) => s + (b - a), 0),
    postResetOobMin: t1 > 60 ? countIn(flat, OOB, 60, t1) : 0,
    episodes: merged.map(([a, b], i) => ({
      index: i + 1,
      startMin: a,
      endMin: b,
      rawMin: b - a,
      activeMin: countIn(flat, OOB, a, b),
    })),
    episodesRaw: oobSpans.length,
    episodesMerged: merged.length,
    // State facts come from the budget machine; a concurrent pause must not
    // mask the fact that the budget itself was exhausted. Loss math, above,
    // uses the flattened track instead.
    firstOobMin: oobSpans.length ? oobSpans[0][0] : null,
    lastRecoveryMin:
      budgetSpans[budgetSpans.length - 1][0] === 'Out of budget' || !oobSpans.length
        ? null
        : oobSpans[oobSpans.length - 1][1],
    openedOob: budgetSpans[0][0] === 'Out of budget',
    closedOob: budgetSpans[budgetSpans.length - 1][0] === 'Out of budget',
    hourlyOob: hourly(OOB),
    hourlyPaused: hourly(PAUSED),
    hourlyNa: hourly(NA),
    track,
    budget: budgetTimeline(
      events.filter((e) => e.machine === 'budget_amount'),
      events.filter((e) => e.machine === 'budget_rule'),
      t0,
      t1,
    ),
    chainBreaks: breaks,
    oobUncertaintyMin: breaks.reduce((s, b) => s + b.ambiguityMin, 0) / 2,
    confidence: naMin > 0 ? 'partial_day' : breaks.length ? 'repaired' : 'clean',
    severity: 0,
    diagnosis: '',
    lost: null,
    perf: null,
  };
}

/** Score every campaign in every day present in the event stream. */
export function scoreAll(events: Event[], mergeGapMin = DEFAULT_MERGE_GAP_MIN): CampaignDay[] {
  const byDay = new Map<string, Map<string, Event[]>>();
  for (const e of events) {
    let campaigns = byDay.get(e.dateKey);
    if (!campaigns) byDay.set(e.dateKey, (campaigns = new Map()));
    const list = campaigns.get(e.campaign);
    if (list) list.push(e);
    else campaigns.set(e.campaign, [e]);
  }

  const results: CampaignDay[] = [];
  for (const dateKey of [...byDay.keys()].sort()) {
    const campaigns = byDay.get(dateKey)!;
    // If the export was cut short (a "Today" pull), do not score the
    // unobserved remainder of the day as if it were in budget.
    let lastSeen = 0;
    for (const evs of campaigns.values()) {
      for (const e of evs) if (e.minute > lastSeen) lastSeen = e.minute;
    }
    const dayEnd = lastSeen >= MINUTES_PER_DAY - 1 ? MINUTES_PER_DAY : lastSeen + 1;
    for (const [campaign, evs] of campaigns) {
      const day = scoreCampaignDay(campaign, dateKey, evs, dayEnd, mergeGapMin);
      if (day !== null) results.push(day);
    }
  }
  return results;
}

/** Structural checks that make chart/table disagreement impossible. */
export function checkInvariants(days: CampaignDay[]): string[] {
  const problems: string[] = [];
  for (const d of days) {
    const total = d.inMin + d.oobMin + d.pausedMin + d.naMin;
    if (total !== MINUTES_PER_DAY) {
      problems.push(`${d.campaign} [${d.dateKey}]: minutes sum to ${total}, not 1440`);
    }
    if (d.episodes.reduce((s, e) => s + e.activeMin, 0) !== d.oobMin) {
      problems.push(`${d.campaign} [${d.dateKey}]: episode minutes != out-of-budget minutes`);
    }
    if (d.hourlyOob.reduce((s, v) => s + v, 0) !== d.oobMin) {
      problems.push(`${d.campaign} [${d.dateKey}]: hourly buckets != out-of-budget minutes`);
    }
    for (let i = 1; i < d.track.length; i++) {
      if (d.track[i][1] !== d.track[i - 1][2]) {
        problems.push(`${d.campaign} [${d.dateKey}]: timeline has a gap or overlap`);
        break;
      }
    }
  }
  return problems;
}

File: ppcbudget/scoring.py
"""Reconstruct each campaign's budget state timeline from change events.

The export lists only *changes*, so the state between two rows has to be
inferred. Three things make that non-obvious:

1. Rows are written newest-first, so rows sharing a minute are also
   newest-first and must be reversed before walking the machine. Sorting on
   timestamp alone silently preserves the wrong intra-minute order -- on the
   reference file that costs 64 campaign-hours and manufactures 14 phantom
   inconsistencies.
2. Delivery state (Paused) overlays budget state. A paused campaign forgoes
   nothing to its budget, so paused minutes are excluded from the loss-eligible
   total and from the in-budget denominator that sets the spend rate.
3. De-duplication during export can drop an intermediate transition, leaving a
   row whose `From` disagrees with the running state. We trust the row over the
   inference and place the implied transition at the midpoint of the gap.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from .ingest import BUDGET_STATES, Event, parse_money

IN, OOB, PAUSED, NA = 0, 1, 2, 3
STATE_NAMES = {IN: "in", OOB: "out_of_budget", PAUSED: "paused", NA: "not_eligible"}

MINUTES_PER_DAY = 1440
DEFAULT_MERGE_GAP_MIN = 5


@dataclass(slots=True)
class ChainBreak:
    at_min: int
    expected_from: str
    saw_from: str
    ambiguity_min: int


@dataclass(slots=True)
class Episode:
    index: int
    start_min: int
    end_min: int
    raw_min: int
    active_min: int  # raw minus any overlapping pause


@dataclass(slots=True)
class BudgetInfo:
    value: float | None = None
    source: str = "unknown"  # daily_budget_event | budget_rule | perf_report | unknown
    time_weighted: float | None = None
    changes: int = 0


@dataclass(slots=True)
class CampaignDay:
    campaign: str
    date_key: str
    t0: int
    t1: int

    in_min: int = 0
    oob_min: int = 0  # loss-eligible: pause and out-of-window already removed
    paused_min: int = 0
    na_min: int = 0
    oob_min_raw: int = 0  # what the budget machine alone says
    post_reset_oob_min: int = 0

    episodes: list[Episode] = field(default_factory=list)
    episodes_raw: int = 0
    episodes_merged: int = 0

    first_oob_min: int | None = None
    last_recovery_min: int | None = None
    opened_oob: bool = False
    closed_oob: bool = False

    hourly_oob: list[int] = field(default_factory=lambda: [0] * 24)
    hourly_paused: list[int] = field(default_factory=lambda: [0] * 24)
    hourly_na: list[int] = field(default_factory=lambda: [0] * 24)
    track: list[tuple[int, int, int]] = field(default_factory=list)  # (state, start, end)

    budget: BudgetInfo = field(default_factory=BudgetInfo)
    chain_breaks: list[ChainBreak] = field(default_factory=list)
    oob_uncertainty_min: float = 0.0
    confidence: str = "clean"  # clean | repaired | partial_day

    # filled in by metrics.py / perfjoin.py
    severity: float = 0.0
    diagnosis: str = ""
    lost: dict | None = None
    perf: object | None = None

    @property
    def eligible_min(self) -> int:
        return self.t1 - self.t0

    @property
    def active_min(self) -> int:
        """Eligible minutes where the campaign could actually have spent."""
        return self.eligible_min - self.paused_min

    @property
    def oob_share(self) -> float:
        return self.oob_min / self.active_min if self.active_min > 0 else 0.0

    @property
    def in_hours(self) -> float:
        return self.in_min / 60

    @property
    def oob_hours(self) -> float:
        return self.oob_min / 60

    @property
    def paused_hours(self) -> float:
        return self.paused_min / 60


def sort_key(e: Event) -> tuple[int, int]:
    """Chronological order. Descending source index reverses the newest-first file."""
    return (e.minute, -e.source_index)


def _walk(events: list[Event], t0: int, t1: int, initial: str):
    """Turn a two-state event stream into contiguous spans over [t0, t1)."""
    spans: list[tuple[str, int, int]] = []
    breaks: list[ChainBreak] = []
    state, prev = initial, t0

    for e in events:
        m = e.minute if e.minute > prev else prev
        if e.from_val != state:
            # A transition went missing. The row is observation, the running
            # state is inference, so trust the row and split the difference.
            mid = (prev + m) // 2
            if mid > prev:
                spans.append((state, prev, mid))
            if m > mid:
                spans.append((e.from_val, mid, m))
            breaks.append(ChainBreak(m, state, e.from_val, m - prev))
        elif m > prev:
            spans.append((state, prev, m))
        state, prev = e.to_val, m

    if t1 > prev:
        spans.append((state, prev, t1))
    return spans, breaks


def _budget_timeline(events: list[Event], rule_events: list[Event],
                     t0: int, t1: int) -> BudgetInfo:
    """Recover the daily budget, which can change during the day."""
    amounts = sorted(events, key=sort_key)
    if amounts:
        opening = amounts[0].from_num
        spans: list[tuple[float | None, int, int]] = []
        value, prev = opening, t0
        for e in amounts:
            m = max(e.minute, prev)
            if m > prev:
                spans.append((value, prev, m))
            value, prev = e.to_num, m
        if t1 > prev:
            spans.append((value, prev, t1))

        weighted = sum(v * (b - a) for v, a, b in spans if v is not None)
        covered = sum(b - a for v, a, b in spans if v is not None)
        return BudgetInfo(
            value=amounts[-1].to_num,
            source="daily_budget_event",
            time_weighted=(weighted / covered) if covered else None,
            changes=len(amounts),
        )

    for e in rule_events:
        amount = parse_money(e.from_val, e.to_val)
        if amount is not None:
            return BudgetInfo(value=amount, source="budget_rule", time_weighted=amount)

    return BudgetInfo()


def score_campaign_day(campaign: str, date_key: str, events: list[Event],
                       day_end_min: int = MINUTES_PER_DAY,
                       merge_gap_min: int = DEFAULT_MERGE_GAP_MIN) -> CampaignDay | None:
    """Score one campaign for one day. Returns None if budget state is unknowable."""
    budget_events = sorted((e for e in events if e.machine == "budget"), key=sort_key)
    if not budget_events:
        return None  # never impute a state we did not observe

    delivery_events = sorted((e for e in events if e.machine == "delivery"), key=sort_key)
    created = [e for e in events if e.machine == "created"]

    # Eligible window. A campaign created at 07:02 is scored over the remaining
    # 16.97h, not a full day -- but only if no budget event precedes creation.
    t1 = min(day_end_min, MINUTES_PER_DAY)
    t0 = 0
    if created:
        birth = min(e.minute for e in created)
        if birth <= budget_events[0].minute and birth < t1:
            t0 = birth
    budget_events = [e for e in budget_events if e.minute >= t0]
    if not budget_events:
        return None
    delivery_events = [e for e in delivery_events if e.minute >= t0]

    day = CampaignDay(campaign=campaign, date_key=date_key, t0=t0, t1=t1)

    budget_spans, breaks = _walk(budget_events, t0, t1, budget_events[0].from_val)
    day.chain_breaks = breaks
    day.oob_uncertainty_min = sum(b.ambiguity_min for b in breaks) / 2

    delivery_initial = delivery_events[0].from_val if delivery_events else "Delivering"
    delivery_spans, _ = _walk(delivery_events, t0, t1, delivery_initial)

    # Flatten: not-eligible > paused > out of budget > in budget.
    flat = bytearray([NA]) * MINUTES_PER_DAY
    for state, a, b in budget_spans:
        flat[a:b] = bytes([OOB if state == "Out of budget" else IN]) * (b - a)
    for state, a, b in delivery_spans:
        if state == "Paused":
            flat[a:b] = bytes([PAUSED]) * (b - a)

    day.in_min = flat.count(IN)
    day.oob_min = flat.count(OOB)
    day.paused_min = flat.count(PAUSED)
    day.na_min = flat.count(NA)
    day.post_reset_oob_min = flat[60:t1].count(OOB) if t1 > 60 else 0
    day.hourly_oob = [flat[h * 60:(h + 1) * 60].count(OOB) for h in range(24)]
    day.hourly_paused = [flat[h * 60:(h + 1) * 60].count(PAUSED) for h in range(24)]
    day.hourly_na = [flat[h * 60:(h + 1) * 60].count(NA) for h in range(24)]

    oob_spans = [(a, b) for s, a, b in budget_spans if s == "Out of budget"]
    day.oob_min_raw = sum(b - a for a, b in oob_spans)
    day.episodes_raw = len(oob_spans)

    # Amazon can release a sliver of budget that is consumed within the same
    # minute, producing zero-length recoveries. Those are pacing noise, not
    # genuine outages, so also report a merged count.
    merged: list[tuple[int, int]] = []
    for a, b in oob_spans:
        if merged and a - merged[-1][1] < merge_gap_min:
            merged[-1] = (merged[-1][0], b)
        else:
            merged.append((a, b))
    day.episodes_merged = len(merged)
    day.episodes = [
        Episode(i + 1, a, b, b - a, flat[a:b].count(OOB))
        for i, (a, b) in enumerate(merged)
    ]

    # State facts come from the budget machine; a concurrent pause must not
    # mask the fact that the budget itself was exhausted. Loss math, above,
    # uses the flattened track instead.
    day.opened_oob = budget_spans[0][0] == "Out of budget"
    day.closed_oob = budget_spans[-1][0] == "Out of budget"
    day.first_oob_min = oob_spans[0][0] if oob_spans else None
    day.last_recovery_min = None if day.closed_oob or not oob_spans else oob_spans[-1][1]

    day.budget = _budget_timeline(
        [e for e in events if e.machine == "budget_amount"],
        [e for e in events if e.machine == "budget_rule"],
        t0, t1,
    )

    # Run-length encode for the report; spans are contiguous and tile the day.
    track: list[tuple[int, int, int]] = []
    run_state, run_start = flat[0], 0
    for m in range(1, MINUTES_PER_DAY):
        if flat[m] != run_state:
            track.append((run_state, run_start, m))
            run_state, run_start = flat[m], m
    track.append((run_state, run_start, MINUTES_PER_DAY))
    day.track = track

    if breaks:
        day.confidence = "repaired"
    if day.na_min > 0:
        day.confidence = "partial_day"
    return day


def score_all(events: list[Event], merge_gap_min: int = DEFAULT_MERGE_GAP_MIN) -> list[CampaignDay]:
    """Score every campaign in every day present in the event stream."""
    by_day: dict[str, dict[str, list[Event]]] = {}
    for e in events:
        by_day.setdefault(e.date_key, {}).setdefault(e.campaign, []).append(e)

    results: list[CampaignDay] = []
    for date_key in sorted(by_day):
        campaigns = by_day[date_key]
        # If the export was cut short (a "Today" pull), do not score the
        # unobserved remainder of the day as if it were in budget.
        last_seen = max(e.minute for evs in campaigns.values() for e in evs)
        day_end = MINUTES_PER_DAY if last_seen >= MINUTES_PER_DAY - 1 else last_seen + 1
        for campaign, evs in campaigns.items():
            day = score_campaign_day(campaign, date_key, evs, day_end, merge_gap_min)
            if day is not None:
                results.append(day)
    return results


def check_invariants(days: list[CampaignDay]) -> list[str]:
    """Structural checks that make chart/table disagreement impossible."""
    problems: list[str] = []
    for d in days:
        total = d.in_min + d.oob_min + d.paused_min + d.na_min
        if total != MINUTES_PER_DAY:
            problems.append(f"{d.campaign} [{d.date_key}]: minutes sum to {total}, not 1440")
        if sum(e.active_min for e in d.episodes) != d.oob_min:
            problems.append(f"{d.campaign} [{d.date_key}]: episode minutes != out-of-budget minutes")
        if sum(d.hourly_oob) != d.oob_min:
            problems.append(f"{d.campaign} [{d.date_key}]: hourly buckets != out-of-budget minutes")
        for i in range(1, len(d.track)):
            if d.track[i][1] != d.track[i - 1][2]:
                problems.append(f"{d.campaign} [{d.date_key}]: timeline has a gap or overlap")
                break
    return problems
